### PR TITLE
(CDAP-5791) Fixed not able to run Kafka Spark streaming in unit-test

### PR DESCRIPTION
**Pull request by chtyim**
- Added a unit-test for Kafka Spark Streaming
- Also made improvement on stopping Spark program.

### CODE REVIEW

File: cdap-common/src/main/java/co/cask/cdap/internal/test/AppJarHelper.java [hunk 1]
```diff
@@ -47,23 +47,31 @@ private AppJarHelper() {
...
-    // Exclude all classes that are visible form the system to the program classloader.
-    ApplicationBundler bundler = new ApplicationBundler(new ClassAcceptor() {
       @Override
       public boolean accept(String className, URL classUrl, URL classPathUrl) {
         if (visibleResources.contains(className.replace('.', '/') + ".class")) {
           return false;
         }
-        // Exclude all Spark classes in the app jar.
-        // We keep the Scala classes in case the app is not Spark but uses Scala in unit-test (CDAP-5168)
+        if (className.startsWith("scala.")) {
```
Comment (bdmogal) [inline cdap-common/src/main/java/co/cask/cdap/internal/test/AppJarHelper.java]: why is this needed? scala classes would have been accepted without this check right?

Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/internal/test/AppJarHelper.java]: In here, right. Anyway the AppFabricTestHelper and this AppJarHelper is a mess right now. Have CDAP-5800 to have it fixed.

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeUtils.java [hunk 2]
```diff
@@ -53,33 +61,93 @@
...
+            });
+        }
         return true;
       }
       return defaultFilter.acceptResource(packageName);
     }
+
+    /**
+     * Gets the set of resources information that are from the Spark Streaming Core. It excludes any
+     * Spark streaming extensions, such as Kafka or Flume. They need to be excluded before they are not
```
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeUtils.java]: They need to be excluded since...?

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/submit/DistributedSparkSubmitter.java [hunk 3]
```diff
@@ -28,6 +28,7 @@
 import java.util.Collections;
 import java.util.List;
 import java.util.Map;
+import java.util.concurrent.ExecutorService;
```
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/submit/DistributedSparkSubmitter.java]: unused import?

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeUtils.java (hunk 2 above)
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeUtils.java]: right.

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeUtils.java [hunk 4]
```diff
@@ -53,33 +61,93 @@
   private static final Logger LOG = LoggerFactory.getLogger(SparkRuntimeUtils.class);
 
   // ClassLoader filter
-  private static final FilterClassLoader.Filter SPARK_PROGRAM_CLASS_LOADER_FILTER = new FilterClassLoader.Filter() {
+  @VisibleForTesting
+  public static final FilterClassLoader.Filter SPARK_PROGRAM_CLASS_LOADER_FILTER = new FilterClassLoader.Filter() {
 
     final FilterClassLoader.Filter defaultFilter = FilterClassLoader.defaultFilter();
+    volatile Set<ClassPath.ResourceInfo> sparkStreamingResources;
```
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeUtils.java]: why is this volatile?

Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeUtils.java]: double checked lock.

Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeUtils.java]: perhaps a comment would be useful for later?

File: cdap-spark-core/src/main/scala/co/cask/cdap/app/runtime/spark/DefaultSparkExecutionContext.scala [hunk 5]
```diff
@@ -109,7 +109,7 @@ class DefaultSparkExecutionContext(runtimeContext: SparkRuntimeContext,
     try {
       // If there is a SparkContext, wait for the ApplicationEnd callback
       // This make sure all jobs' transactions are committed/invalidated
-      SparkRuntimeEnv.stop.foreach(sc => applicationEndLatch.await())
+      SparkRuntimeEnv.stop().foreach(sc => applicationEndLatch.await())
```
Comment (bdmogal) [inline cdap-spark-core/src/main/scala/co/cask/cdap/app/runtime/spark/DefaultSparkExecutionContext.scala]: for my information - why is this needed?

File: cdap-spark-core/src/main/scala/co/cask/cdap/app/runtime/spark/SparkMainWrapper.scala [hunk 6]
```diff
@@ -66,31 +84,38 @@ object SparkMainWrapper {
...
-          case cls =>
-            getMainMethod(cls).fold(
-              throw new IllegalArgumentException(userSparkClass.getName
-                + " is not a supported Spark program. It should implement either "
-                + classOf[SparkMain].getName + " or " + classOf[JavaSparkMain].getName
-                + " or has a main method defined")
-            )(
-              _.invoke(null, RuntimeArguments.toPosixArray(runtimeContext.getRuntimeArguments))
-            )
+        val cancelHeartbeat = startHeartbeat(arguments, runtimeContext, () => triggerShutdown)
```
Comment (bdmogal) [inline cdap-spark-core/src/main/scala/co/cask/cdap/app/runtime/spark/SparkMainWrapper.scala]: is the indentation off for this block of code?

File: cdap-spark-core/src/main/scala/co/cask/cdap/app/runtime/spark/SparkRuntimeEnv.scala [hunk 7]
```diff
@@ -153,28 +153,49 @@ object SparkRuntimeEnv {
...
     this.synchronized {
       if (!stopped) {
         stopped = true
+        sc = sparkContext
+        ssc = streamingContext
+      }
+    }
+
+    try {
+      ssc.foreach(context => {
```
Comment (bdmogal) [inline cdap-spark-core/src/main/scala/co/cask/cdap/app/runtime/spark/SparkRuntimeEnv.scala]: what happens when ``ssc`` is ``None``? Or am I reading this wrong?

Comment (bdmogal) [inline cdap-spark-core/src/main/scala/co/cask/cdap/app/runtime/spark/SparkRuntimeEnv.scala]: Oh nvm. Its an ``Option`` 👍

File: cdap-unit-test/pom.xml [hunk 8]
```diff
@@ -123,6 +123,27 @@
       <version>${project.version}</version>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>co.cask.cdap</groupId>
+      <artifactId>cdap-data-fabric</artifactId>
+      <version>${project.version}</version>
+      <type>test-jar</type>
```
Comment (bdmogal) [inline cdap-unit-test/pom.xml]: for ``KafkaTester``? If so, should we move that class to a better place? Not for this PR though.

File: cdap-unit-test/src/main/java/co/cask/cdap/test/UnitTestManager.java [hunk 9]
```diff
@@ -86,6 +91,24 @@
 public class UnitTestManager implements TestManager {
 
   private static final Gson GSON = new Gson();
+  private static final ClassAcceptor CLASS_ACCEPTOR = new ClassAcceptor() {
```
Comment (bdmogal) [inline cdap-unit-test/src/main/java/co/cask/cdap/test/UnitTestManager.java]: should we define this in the interface (``TestManager``) if it is used by both ``UnitTestManager`` and ``IntegrationTestManager``?

File: cdap-unit-test/src/main/java/co/cask/cdap/test/UnitTestManager.java [hunk 10]
```diff
@@ -148,21 +171,24 @@ public ApplicationManager deployApplication(Id.Namespace namespace,
...
         configObject = (Config) TypeToken.of(configType).getRawType().newInstance();
       }
 
       Application app = applicationClz.newInstance();
       MockAppConfigurer configurer = new MockAppConfigurer(app);
       app.configure(configurer, new DefaultApplicationContext<>(configObject));
-      appFabricClient.deployApplication(namespace, applicationClz, appConfig, bundleEmbeddedJars);
       ApplicationId applicationId = new ApplicationId(namespace.getId(), configurer.getName());
+
+      ArtifactSummary artifactSummy = new ArtifactSummary(artifactId.getArtifact(), artifactId.getVersion());
```
Comment (bdmogal) [inline cdap-unit-test/src/main/java/co/cask/cdap/test/UnitTestManager.java]: typo - artifactSummary

File: cdap-unit-test/src/test/java/co/cask/cdap/test/app/SparkCSVToSpaceProgram.java [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+import java.io.PrintWriter;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Main class for the Spark program to convert comma separated file into space separated file.
+ */
+
```
Comment (bdmogal) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/SparkCSVToSpaceProgram.java]: nit: extra newline

File: cdap-unit-test/src/test/java/co/cask/cdap/spark/SparkStreamingTestRun.java [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+import org.apache.twill.kafka.client.KafkaPublisher;
+import org.junit.ClassRule;
+import org.junit.Test;
+
+import java.util.Iterator;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+
+/**
+ *
```
Comment (bdmogal) [inline cdap-unit-test/src/test/java/co/cask/cdap/spark/SparkStreamingTestRun.java]: nit: empty javadoc.

File: cdap-client-tests/src/test/java/co/cask/cdap/client/app/NoOpSparkProgram.java [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+import co.cask.cdap.api.spark.JavaSparkExecutionContext;
+import co.cask.cdap.api.spark.JavaSparkMain;
+import org.apache.spark.api.java.JavaPairRDD;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ *
```
Comment (bdmogal) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/app/NoOpSparkProgram.java]: nit: empty javadoc

File: cdap-unit-test/src/test/java/co/cask/cdap/test/app/SparkCSVToSpaceProgram.java [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+    sec.execute(new TxRunnable() {
+      @Override
+      public void run(DatasetContext context) throws Exception {
+        Map<String, String> args = sec.getRuntimeArguments();
+        String outputPath = args.get("output.path");
+
+        Map<String, String> fileSetArgs = new HashMap<>();
+        FileSetArguments.setOutputPath(fileSetArgs, outputPath);
+
+
```
Comment (bdmogal) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/SparkCSVToSpaceProgram.java]: nit: extra newline

File: cdap-unit-test/src/test/java/co/cask/cdap/test/app/SparkCSVToSpaceProgram.java [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+import java.io.PrintWriter;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Main class for the Spark program to convert comma separated file into space separated file.
+ */
+
+public final class SparkCSVToSpaceProgram extends AbstractSpark implements JavaSparkMain {
```
Comment (bdmogal) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/SparkCSVToSpaceProgram.java]: also, is it necessary to move spark programs to their own class? I see that pattern here a couple of times.

File: cdap-spark-core/src/main/scala/co/cask/cdap/app/runtime/spark/DefaultSparkExecutionContext.scala (hunk 5 above)
Comment (chtyim) [inline cdap-spark-core/src/main/scala/co/cask/cdap/app/runtime/spark/DefaultSparkExecutionContext.scala]: The `()`? Because the `stop` method now has a default argument and in Scala, you can only omit the `()` if a method has no argument.

**Comment by bdmogal** [comment]
@chtyim just a few minor comments. overall the code changes LGTM

File: cdap-spark-core/src/main/scala/co/cask/cdap/app/runtime/spark/DefaultSparkExecutionContext.scala (hunk 5 above)
Comment (bdmogal) [inline cdap-spark-core/src/main/scala/co/cask/cdap/app/runtime/spark/DefaultSparkExecutionContext.scala]: Ah - right!

File: cdap-spark-core/src/main/scala/co/cask/cdap/app/runtime/spark/SparkMainWrapper.scala (hunk 6 above)
Comment (chtyim) [inline cdap-spark-core/src/main/scala/co/cask/cdap/app/runtime/spark/SparkMainWrapper.scala]: Don't think so. It's github messed up.

File: cdap-unit-test/pom.xml (hunk 8 above)
Comment (chtyim) [inline cdap-unit-test/pom.xml]: Probably. The unit-tests and dependencies are pretty messy now.

File: cdap-unit-test/src/main/java/co/cask/cdap/test/UnitTestManager.java (hunk 9 above)
Comment (chtyim) [inline cdap-unit-test/src/main/java/co/cask/cdap/test/UnitTestManager.java]: For now I keep them separated in case they need to behaves differently (on unit-test all classes are from system classpath vs integrated test may not).

File: cdap-unit-test/src/test/java/co/cask/cdap/spark/SparkStreamingTestRun.java (hunk 12 above)
Comment (chtyim) [inline cdap-unit-test/src/test/java/co/cask/cdap/spark/SparkStreamingTestRun.java]: It's test case.

File: cdap-unit-test/src/test/java/co/cask/cdap/test/app/SparkCSVToSpaceProgram.java (hunk 15 above)
Comment (chtyim) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/SparkCSVToSpaceProgram.java]: For cases, yes. The reason is because if a Spark class is defined as a inner class, it may cause `JavaSparkMain` class not found issue for other inner classes, depending on where the Spark class is defined (it is related to where it appears in the bytecode of the enclosing class). Moving it out will resolve the issue.

File: cdap-unit-test/src/test/java/co/cask/cdap/test/app/SparkCSVToSpaceProgram.java (hunk 14 above)
Comment (chtyim) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/SparkCSVToSpaceProgram.java]: These are from existing code. Anyway will fix it.

File: cdap-client-tests/src/test/java/co/cask/cdap/client/app/NoOpSparkProgram.java (hunk 13 above)
Comment (chtyim) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/app/NoOpSparkProgram.java]: It was like that before and it's a test class.

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeUtils.java (hunk 4 above)
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeUtils.java]: It's a common pattern. Do we need comment??

**Comment by chtyim** [comment]
Build passed: http://builds.cask.co/browse/CDAP-DUT4053

The last commit was to address comments, which involves removing a unless `if`, code comment fixes, formatting and a variable rename. Not going to wait for another build.
